cpu/atmega8/Makefile.features

```diff
@@ -1 +1,3 @@
 include $(RIOTCPU)/atmega_common/Makefile.features
+
+FEATURES_BLACKLIST += periph_cpuid
```

Comment:
Is `boot_signature_byte_get()` not implemented on this CPU?


```patch
--- a/cpu/atmega_common/Makefile.features
+++ b/cpu/atmega_common/Makefile.features
@@ -6,7 +6,6 @@ include $(RIOTCPU)/avr8_common/Makefile.features
 FEATURES_PROVIDED += cpu_core_atmega
 FEATURES_PROVIDED += atmega_pcint0
 FEATURES_PROVIDED += dbgpin
-FEATURES_PROVIDED += periph_cpuid
 FEATURES_PROVIDED += periph_eeprom
 FEATURES_PROVIDED += periph_gpio periph_gpio_irq
 FEATURES_PROVIDED += periph_gpio_ll
@@ -20,6 +19,9 @@ FEATURES_PROVIDED += periph_timer_periodic
 FEATURES_PROVIDED += periph_rtt_overflow
 FEATURES_PROVIDED += periph_wdt
 FEATURES_PROVIDED += puf_sram
+ifneq (atmega8, $(CPU_FAM))
+  FEATURES_PROVIDED += periph_cpuid
+endif
 
 FEATURES_CONFLICT += periph_rtc:periph_rtt
 FEATURES_CONFLICT_MSG += "On ATmega, the RTC and RTT use to the same hardware timer."
```

Reply:
In avr/boot.h:
```c
#ifndef __DOXYGEN__
#define __BOOT_SIGROW_READ (_BV(__SPM_ENABLE) | _BV(SIGRD))
#endif
/** \ingroup avr_boot
    \def boot_signature_byte_get(address)

    Read the Signature Row byte at \c address.  For some MCU types,
    this function can also retrieve the factory-stored oscillator
    calibration bytes.

    Parameter \c address can be 0-0x1f as documented by the datasheet.
    \note The values are MCU type dependent.
*/

#define boot_signature_byte_get(addr) \
(__extension__({                      \
      uint8_t __result;                         \
      __asm__ __volatile__                      \
      (                                         \
        "sts %1, %2\n\t"                        \
        "lpm %0, Z" "\n\t"                      \
        : "=r" (__result)                       \
        : "i" (_SFR_MEM_ADDR(__SPM_REG)),       \
          "r" ((uint8_t)(__BOOT_SIGROW_READ)),  \
          "z" ((uint16_t)(addr))                \
      );                                        \
      __result;                                 \
}))
```
In avr/iom8.h:
```c
/* SPMCR */
#define SPMIE	7
#define RWWSB	6
/* bit 5 reserved */
#define RWWSRE	4
#define BLBSET	3
#define PGWRT	2
#define PGERS	1
#define SPMEN	0
```
This results in:
```
/home/hugues/github/boards_atmega8/RIOT/cpu/atmega_common/periph/cpuid.c:31:14: error: ‘SIGRD’ undeclared (first use in this function)
     out[0] = boot_signature_byte_get(0x00);     /* Atmel manufacturer ID (0x1E) */
```
> The problem was the lack of definition of the SIGRD bit of the SPMCR register in avr/iom8.h. According to the datasheet, bit 5 of this register is reserved, but this is also the case for the atmega328p, although it is defined in avr/iom328p.h and the driver works correctly for it. So I tried to define it in the driver, which enabled test/periph/cpuid to be compiled for atmega8. However, unlike atmega328p, the values returned were incorrect (0xbc 0xe7 0xe6 instead of 0x1e 0x93 0x07 for atmega8).

Reply:
Moreover there is no _Reading the Signature Row from Software_ section in the datasheet so I think it is simply not possible.

Reply:
I still had to add `CPU_FAM = atmega8` to `atmega8/Makefile.features` for it to work. Now `tests/build_system/kconfig_features` fails...

Reply:
I think you also need

```patch
--- a/cpu/atmega_common/Kconfig
+++ b/cpu/atmega_common/Kconfig
@@ -15,7 +15,7 @@ config CPU_COMMON_ATMEGA
     select HAS_CPU_CORE_ATMEGA
     select HAS_ATMEGA_PCINT0
     select HAS_DBGPIN
-    select HAS_PERIPH_CPUID
+    select HAS_PERIPH_CPUID if !CPU_FAM_ATMEGA8
     select HAS_PERIPH_EEPROM
     select HAS_PERIPH_GPIO
     select HAS_PERIPH_GPIO_IRQ
```

Reply:
Thanks. I'm not comfortable with the buildsystem. Is `atmega8/Makefile.fearures` the right place to define CPU_FAM?

Reply:
Yes that's also where other CPUs do it.